L1Trigger/L1THGCalUtilities/plugins/CaloTruthCellsProducer.cc

```diff
@@ -0,0 +1,397 @@
+      continue;
+    if (genIndex >= int(orderedCaloRefs.size()))
+      orderedCaloRefs.resize(genIndex + 1);
+
+    orderedCaloRefs[genIndex] = ref;
+  }
+
+  auto outMap(std::make_unique<CaloToCellsMap>(caloParticlesHandle, triggerCellsHandle));
+  std::unique_ptr<l1t::HGCalTriggerCellBxCollection> outCollection;
+  if (makeCellsCollection_)
+    outCollection.reset(new l1t::HGCalTriggerCellBxCollection);
+
+  int bx(triggerCells.getFirstBX());
+  unsigned bxOffset(0);
+
+  typedef edm::Ptr<l1t::HGCalTriggerCell> TriggerCellPtr;
+  typedef edm::Ptr<l1t::HGCalCluster> ClusterPtr;
+
+  // ClusteringDummyImpl only considers BX 0, so we dump all cells to one vector
+  std::vector<TriggerCellPtr> triggerCellPtrs;
+
+  // loop through all bunch crossings
+  for (unsigned iC(0); iC != triggerCells.size(); ++iC) {
+    if (iC >= bxOffset + triggerCells.size(bx)) {
+      bxOffset += triggerCells.size(bx);
+      bx += 1;
+    }
+
+    auto& cell(triggerCells[iC]);
+
+    auto mapElem(tcToCalo.find(cell.detId()));
+    if (mapElem == tcToCalo.end())
+      continue;
+
+    edm::Ref<l1t::HGCalTriggerCellBxCollection> ref(triggerCellsHandle, iC);
+
+    outMap->insert(mapElem->second, ref);
+
+    if (makeCellsCollection_)
+      outCollection->push_back(bx, cell);
+
+    triggerCellPtrs.emplace_back(triggerCellsHandle, iC);
+  }
+
+  _event.put(std::move(outMap));
+  if (makeCellsCollection_)
+    _event.put(std::move(outCollection), "HGCalVFEProcessorSums");
+
+  auto outClusters(std::make_unique<l1t::HGCalClusterBxCollection>());
+
+  auto sortCellPtrs([](TriggerCellPtr const& lhs, TriggerCellPtr const& rhs)->bool {
+      return lhs->mipPt() > rhs->mipPt();
+    });
+
+  std::sort(triggerCellPtrs.begin(), triggerCellPtrs.end(), sortCellPtrs);
+  dummyClustering_.clusterizeDummy(triggerCellPtrs, *outClusters);
+
+  std::map<CaloParticleRef, std::vector<unsigned>> caloToClusterIndices;
+  for (unsigned iC(0); iC != outClusters->size(); ++iC) {
+    auto& cluster((*outClusters)[iC]);
+    // cluster detId and cell detId are identical
+    auto caloRef(tcToCalo.at(cluster.detId()));
+    caloToClusterIndices[caloRef].push_back(iC);
+  }
+
+  auto&& clustersHandle(_event.put(std::move(outClusters), "HGCalBackendLayer1Processor2DClustering"));
+
+  auto outMulticlusters(std::make_unique<l1t::HGCalMulticlusterBxCollection>());
+
+  for (auto& ref : orderedCaloRefs) {
+    if (ref.isNull()) // shouldn't happen
+      continue;
+
+    auto& caloParticle(*ref);
+
+    l1t::HGCalTriggerCell dummyCell;
+    dummyCell.setDetId(caloParticle.g4Tracks().at(0).genpartIndex() - 1);
+    TriggerCellPtr tcptr(&dummyCell, 0);
+
+    l1t::HGCalCluster dummyCluster(tcptr);
+    ClusterPtr clptr(&dummyCluster, 0);
+
+    l1t::HGCalMulticluster multicluster(clptr); // set the detid
+    
+    for (unsigned iC : caloToClusterIndices[ref]) {
+      ClusterPtr clPtr(clustersHandle, iC);
+      multicluster.addConstituent(clPtr, true, 1.);
```

Comment:
The fraction shouldn't be 1, but should correspond to the fraction computed from the input collection of sim clusters

Reply:
This relates to the energy fraction discussion above. I think the only way to have a fully "correct" energy fraction is to start from truth-matched digis and form another collection of trigger cells. Gianluca and I discussed this in the past and decided that would be an overkill, and opted instead to accept some (likely small) inaccuracies arising from energy mixing from pileup into the trigger cells. We can discuss this further though.

Reply:
I don't think another collection of trigger cells is needed. I think we just need to have access to the simhits (see my comment above) to compute the TC fractions. The cluster `addConstituent` method can then take fractions.

Reply:
I kept the fraction here as 1 because (as mentioned above) the code is based on the assumption that each truth TC (after subtracting the PU energy using hitsAndFractions) belongs to one CaloParticle. Under this assumption, each Cluster will also belong to one CaloParticle = one Multicluster, so this fraction is always 1.

Reply:
As mentioned above I don't think this assumption is valid